packages/data-context/src/sources/BrowserDataSource.ts

```diff
@@ -33,6 +42,41 @@ export interface BrowserApiShape {
 export class BrowserDataSource {
   constructor (private ctx: DataContext) {}
 
+  /**
+   * Gets the browsers from the machine and the project config
+   */
+  allBrowsers () {
+    if (!this.ctx.coreData.allBrowsers) {
+      const p = this.ctx.project.getConfig()
+      const machineBrowsers = this.machineBrowsers()
+
+      this.ctx.coreData.allBrowsers = Promise.all([p, machineBrowsers]).then(async ([cfg, m]) => {
+        if (!cfg.browsers) return m
+
+        const userBrowsers = cfg.browsers.map<FoundBrowser | undefined>((b) => {
+          if (_.includes(_.map(m, getBrowserKey), getBrowserKey(b))) return
+
+          return {
+            ...b,
+            majorVersion: String(b.majorVersion),
+            custom: true,
+          }
+        }).filter<FoundBrowser>((b): b is FoundBrowser => b != null) || []
+
+        return _.concat(m, userBrowsers)
+      }).catch((e) => {
+        this.ctx.update((coreData) => {
+          coreData.allBrowsers = null
+          coreData.diagnostics.error = e
```

Comment:
Nice job figuring out where/how to propagate this error, the data-context is pretty confusing. Did you actually verify this error manifests in the UI/CLI?

I'm also not entirely clear on the condition that would lead to this - is there one, or did you just put it here for good measure?

Reply:
Gonna need a quick test for both cases, I don't think we have much coverage around detecting the browsers in general, I could pick this up.

Reply:
I don't think the new logic will likely error. `machineBrowsers()` re-throws so that'll be caught here but really you're just getting two copies of the same error then. Lemme experiment a bit to see if I can break it. ;)

Reply:
Looks like the config validation logic would catch the cases in which this code would fail so I can remove the error handling.